### PR TITLE
Add separate hookable tracing module

### PR DESCRIPTION
**Pull request by andy31415**
Looking to solve the difficulty in figuring out what applications are doing (e.g. java matter controller tests fail occasionally and I still do not know why).

We do have the infrastructure, however it is rarely actually used:
  - pw_tracing is available but rarely used and overrides all other tracing
  - tracing hooks exist if not using pigweed, but no actual real implementation in the SDK (I plan to add a perfetto one)
  - data logging is using separate and non-typesafe paths (void * passing around, with string event compares) and outputs data that does not seem useful to debug (encrypted data). It is also caled `examples/tracing` which is not the same as `src/trace` but rather uses `MessageTracing.h`

This just adds a generic tracing backend *with some unit tests*.
Intent is to have followups for:
  - Provide a backend with similar functionality to `examples/tracing` (TraceStream/TraceLog/TraceFile/...)
  - Provide a perfetto-based backend
  - Provide a pw_trace-based backend
  - Convert existing tracing and data logging (for messages) to the new infrastructure
  - Add additional logging (we are missing DNSSD data logging currently)
  - Adjust formats as needed (CSG asked for more spec 4.4 Message format data, a human-friendly TLV decoding would also be great)
  
 Overall goal is to be able to figure out "what went wrong in a failed commissioning" (as an example) within minutes of having a log. Today I have seen people knowing the SDK well spending hours in log reading + wireshark captures to figure out things and I believe we can improve on that.

### CODE REVIEW

File: src/tracing/registry.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    gTracingBackends.Remove(&backend);
+}
+
+#ifdef MATTER_TRACING_ENABLED
+
+namespace Internal {
+
+void Begin(::chip::Tracing::Scope scope)
+{
+    for (auto backend = gTracingBackends.begin(); backend != gTracingBackends.end(); backend++)
```
Comment (bzbarsky-apple) [inline src/tracing/registry.cpp:48]: Any reason not to `for (auto & backend : gTracingBackends)` ?

File: src/tracing/registry.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+namespace chip {
+namespace Tracing {
+
+/// Registers a tracing backend to receive trace and logging data
+/// Until it is unregistered
+///
+/// All tracing backends MUST be unregistered before the application
+/// exits. Consider using [ScopedRegistration]
+///
+/// MUST be called with chip thread lock held (from chip main loop or
```
Comment (bzbarsky-apple) [inline src/tracing/registry.h:30]:
```suggestion
/// MUST be called with Matter thread lock held (from Matter main loop or
```
and similar elsewhere?

File: src/tracing/scopes.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    PASESession_HandlePBKDFParamRequest,
+    PASESession_HandlePBKDFParamResponse,
+    PASESession_Pair,
+    PASESession_SendMsg1,
+    PASESession_SendPBKDFParamRequest,
+    PASESession_SendPBKDFParamResponse,
+    PASESession_SetupSpake2p,
+};
+
+/// An event that happened instant (like a zero sized scope)
```
Comment (bzbarsky-apple) [inline src/tracing/scopes.h:98]:
```suggestion
/// An event that happened at an instant (like a zero sized scope)
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/tracing/registry.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/tracing/registry.cpp:48]: +1

File: src/tracing/scopes.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+/// As a result, tracing scopes in CHIP are from a known list.
+enum class Scope
+{
+    CASESession_EstablishSession,
+    CASESession_HandleSigma1,
+    CASESession_HandleSigma1_and_SendSigma2,
+    CASESession_HandleSigma2,
+    CASESession_HandleSigma2_and_SendSigma3,
+    CASESession_HandleSigma2Resume,
+    CASESession_HandleSigma3,
```
Comment (tcarmelveilleux) [inline src/tracing/scopes.h:34]: Please add an ` = X` to each and start at 1, with 0 being undefined. Will make it easier to look at trace values in logs and correlate.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/tracing/registry.cpp (hunk 1 above)
Comment (andy31415) [reply src/tracing/registry.cpp:48]: This is what I wrote at first and I got compiler errors I belive .... will try it again to check...

Comment (andy31415) [reply src/tracing/registry.cpp:48]: This is what I wrote at first and I got compiler errors I belive .... will try it again to check...

Comment (andy31415) [reply src/tracing/registry.cpp:48]: I had used the begin/end and that fails:

```
2023-06-01 13:19:36 INFO    ../../src/tracing/registry.cpp:48:49: error: cannot bind non-const lvalue reference of type ‘chip::IntrusiveList<chip::Tracing::Backend>::Iterator&’ to an rvalue of type ‘chip::IntrusiveList<chip::Tracing::Backend>::Iterator’
2023-06-01 13:19:36 INFO       48 |     for (auto & backend = gTracingBackends.begin(); backend != gTracingBackends.end(); backend++)
2023-06-01 13:19:36 INFO          |  
```

however your form works. Updating.

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**